Restore build properties

jasontedor opened the pull request:
This commit restores the build properties provided in
org.elasticsearch.Build. This class previously obtained the build hash
and timestamp from a resource es-build.properties that was included in
the jar and produced by the Maven resources plugin. After the switch to
Gradle, the production of this file was lost and these build properties
defaulted to “NA” in all instances.

The most important place that the build hash is used is in the plugin
manager to determine the URL of staging artifacts for plugins.

The build hash is also used in several responses including the /_nodes
response and the response to HTTP GET requests on the root path.

These properties can now be obtained from the jar manifest as they are
currently placed there by the gradle-info plugin. However, only the
short hash is provided. We now read the manifest for these properties
and no longer provide the full hash in responses to HTTP GET requests
on the root path.

nik9000 commented:
Fine by me.

rmuir commented:
do we know what uses these? we also have this information (and more) in the jar metadata.

jasontedor commented:
@rmuir It's used when you hit the root path against an Elasticsearch node. So today if you hit the root path against a node in Elasticsearch cluster built from master you will see

```
    "build_hash" : "${buildNumber}",
    "build_timestamp" : "NA",
```

in the response. This comes from the included resource `es-build.properties` having contents

```
version=${project.version}
hash=${buildNumber}
timestamp=${timestamp}
```

 With this pull request you will see

```
    "build_hash" : "b328e26160b47abe904dfd08e920d31f668b7e57",
    "build_timestamp" : "2015-11-05T17:45:12Z",
```

and `es-build.properties` now looks like

```
#Thu Nov 05 12:45:19 EST 2015
version=3.0.0-SNAPSHOT
timestamp=1446745512970
hash=b328e26160b47abe904dfd08e920d31f668b7e57
```

rmuir commented:
Right, but it could alternatively get it from jar metadata? and then we don't need this file?

rjernst commented:
I agree with @rmuir, we should use the jar manifest if possible.

jasontedor committed:
Restore build properties

This commit restores the build properties provided in
org.elasticsearch.Build. This class previously obtained the build hash
and timestamp from a resource es-build.properties that was included in
the jar and produced by the Maven resources plugin. After the switch to
Gradle, the production of this file was lost and these build properties
defaulted to “NA” in all instances.

The most important place that the build hash is used is in the plugin
manager to determine the URL of staging artifacts for plugins.

The build hash is also used in several responses including the /_nodes
response and the response to HTTP GET requests on the root path.

These properties can now be obtained from the jar manifest as they are
currently placed there by the gradle-info plugin. However, only the
short hash is provided. We now read the manifest for these properties
and no longer provide the full hash in responses to HTTP GET requests
on the root path.

jasontedor commented:
@rmuir @rjernst I've pushed commit 45475fa1cdfc48b84542377edb81cc88b4674e53 reverting the previous changes to the build and added e5b0d38e4ed617d7f7e39ba9e5682c660cfab625 to obtain these properties from the manifest.

rjernst commented on core/src/main/java/org/elasticsearch/Build.java:
```diff
@@ -35,65 +33,51 @@
-                timestamp = ISODateTimeFormat.dateTimeNoMillis().withZone(DateTimeZone.UTC).print(Long.parseLong(gitTimestampRaw));
-            }
-        } catch (Exception e) {
+        String change = "NA";
```
I know this was the same before, but isn't this more "Unknown"?

rjernst commented on core/src/main/java/org/elasticsearch/Build.java:
```diff
@@ -35,65 +33,51 @@
 
-    public String hash() {
-        return hash;
+    Build(String hashShort, String timestamp) {
```
nit: this is normally called the "short hash" so reversing it sounds very odd. :)

rjernst commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoResponse.java:
```diff
@@ -77,7 +77,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
 
             builder.field("version", nodeInfo.getVersion());
-            builder.field("build", nodeInfo.getBuild().hashShort());
+            builder.field("build", nodeInfo.getBuild().change());
```
Maybe output this as "build-hash"?

rjernst commented on core/src/main/java/org/elasticsearch/rest/action/main/RestMainAction.java:
```diff
@@ -77,8 +77,8 @@ public void handleRequest(final RestRequest request, RestChannel channel, final
                 .field("number", version.number())
-                .field("build_hash", Build.CURRENT.hash())
-                .field("build_timestamp", Build.CURRENT.timestamp())
+                .field("build_change", Build.CURRENT.change())
```
Same here, I would call this "build_hash"

rjernst commented:
LGTM

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy:
```diff
@@ -83,6 +83,19 @@ class BuildPlugin implements Plugin<Project> {
     /** Adds additional manifest info to jars */
     static void configureJarManifest(Project project) {
+        project.scminfo {
+            def dotgit = "$project.rootProject.projectDir/.git/"
```
I would normally use the form `${foo}` (add curlies)

rjernst replied:
You could also calulate the hash once and store as an ext prop on the rootProject. See for example above how we only execute globalBuildInfo once.

rjernst replied:
Hrm, also I think we need to do a traversal here, going up from the current project's project dir. We can't assume rootProject is the git root (for example, if elasticsearch is attached to another project with the project-attachment-plugin).

jasontedor replied:
All of these comments are addressed by 95dd9c13ad77bde2e5ad5f3ef4e7669f0be8e51d.

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy:
```diff
@@ -83,6 +83,19 @@ class BuildPlugin implements Plugin<Project> {
     static void configureJarManifest(Project project) {
+        project.scminfo {
+            def dotgit = "$project.rootProject.projectDir/.git/"
+            def head = new File(dotgit + "HEAD").text.split(":")
```
text -> getText('UTF-8')

rjernst replied:
I've been preferring to actually use types vs `def`. I don't think there is any advantage to def, and why not when we know the types anyways? It makes the code easier to read too.

jasontedor replied:
Both of these comments are addressed by 95dd9c13ad77bde2e5ad5f3ef4e7669f0be8e51d.

jasontedor commented:
For posterity, another use that has been discovered is for the paths for staging artifacts in the plugin manager. It's way outside the scope of this issue to change the release process to not use the full hash, and we do have short hash (the first seven characters of the full hash as coming from the gradle-info plugin) collisions for commit objects. As such, we do need to get the full hash somehow which is the point of the [change](https://github.com/jasontedor/elasticsearch/blob/95dd9c13ad77bde2e5ad5f3ef4e7669f0be8e51d/buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy#L88-L90) to `BuildPlugin.groovy`.

jasontedor commented:
The only place that the full hash was used is in the response to HTTP GET requests on the root path. All other places used the short hash. This means that we can safely use the gradle-info plugin (which provides the short hash as the `Change` attribute in the jar manifest) and just break the fact that the response to HTTP GET against the root path included the full hash.

rjernst commented:
Lgtm, thanks Jason
